lib/mauth/client.rb

```diff
@@ -293,55 +320,95 @@ def authentic?(object)
         begin
           authenticate!(object)
           true
-        rescue InauthenticError, MauthNotPresent
+        rescue InauthenticError, MauthNotPresent, MAuth::MissingV2Error
           false
         end
       end
 
-      # raises InauthenticError unless the given object is authentic
+      # raises InauthenticError unless the given object is authentic. Will only
+      # authenticate with v2 if the environment variable V2_ONLY_AUTHENTICATE
+      # is set. Otherwise will authenticate with only the highest protocol version present
       def authenticate!(object)
-        authentication_present!(object)
-        time_valid!(object)
-        token_valid!(object)
-        signature_valid!(object)
-      rescue MauthNotPresent => e
-        logger.warn "mAuth signature not present on #{object.class}. Exception: #{e.message}"
-        raise
-      rescue InauthenticError => e
-        logger.error "mAuth signature authentication failed for #{object.class}. Exception: #{e.message}"
-        raise
-      rescue UnableToAuthenticateError => e
-        logger.error "Unable to authenticate with MAuth for #{object.class}. Exception: #{e.message}"
-        raise
+        if authentication_present_v2?(object)
```

Comment:
Correct me if I'm wrong here, but the "object" being passed to this function should be either a `MAuth::Faraday::Response` or a `MAuth::Rack::Request`?

If that is correct, then I think we could make this logic and error throwing simpler by defining a `version` function on those two classes.

Also, note for future refactor: it would prob be much nicer to have two base classes called `Incoming` and `Outgoing` as they are essentially the inverse with respect to signing vs authenticating for rack and faraday and we could save some duplicate functionality.

Reply:
I put a `#protocol_version` method in the `Signed` module in `request_and_response.rb`. The `Signed` module serves basically as a place to put `Incoming` methods.